File: day15/.ts
type Data = Array<Record<string, string | number>>
type SortBy = string

function drawTable(data: Data, sortBy: SortBy): string {
  if (!data || data.length === 0) return "";

  const keys = Object.keys(data[0]);

  const sortedData = [...data].sort((a, b) => {
    const valA = a[sortBy];
    const valB = b[sortBy];
    if (typeof valA === "number" && typeof valB === "number") {
      return valA - valB;
    }
    return String(valA).localeCompare(String(valB));
  });

  const widths = {};
  for (const key of keys) widths[key] = 0;
  for (const row of sortedData) {
    for (const key of keys) {
      widths[key] = Math.max(widths[key], String(row[key]).length);
    }
  }

  const borderSegments = keys.map(key => "+" + "-".repeat(widths[key] + 2));
  const borderRow = borderSegments.join("") + "+";

  const headerSegments = keys.map((key, i) => {
    const header = String.fromCharCode(65 + i); // Idea taken from Bryan Herrera
    const padding = " ".repeat(widths[key] - header.length + 1);
    return "| " + header + padding;
  });
  const headerRow = headerSegments.join("") + "|";

  const dataRows = sortedData.map(row => {
    const segments = keys.map(key => {
      const value = String(row[key]);
      const totalWidth = widths[key] + 1;
      const padding = " ".repeat(Math.max(0, totalWidth - value.length));
      return "| " + value + padding;
    });
    return segments.join("") + "|";
  }).join("\n");

  return [
    borderRow,
    headerRow,
    borderRow,
    dataRows,
    borderRow
  ].join("\n");
}

File: day15/.py
def draw_table(data: list[dict[str, str | int]], sortBy: str) -> str:
  if not data or len(data) == 0:
    return ""

  keys = list(data[0].keys())

  def sort_key(row):
    val = row[sortBy]
    if isinstance(val, (int, float)):
      return (0, val)
    else:
      return (1, str(val))

  sorted_data = sorted(data, key=sort_key)

  widths = {}
  for key in keys:
    widths[key] = 0

  for row in sorted_data:
    for key in keys:
      widths[key] = max(widths[key], len(str(row[key])))

  border_segments = ["+" + "-" * (widths[key] + 2) for key in keys]
  border_row = "".join(border_segments) + "+"

  header_segments = []
  for i, key in enumerate(keys):
    header = chr(65 + i)  # Idea taken from Bryan Herrera
    padding = " " * (widths[key] - len(header) + 1)
    header_segments.append("| " + header + padding)
  header_row = "".join(header_segments) + "|"

  data_row_list = []
  for row in sorted_data:
    segments = []
    for key in keys:
      value = str(row[key])
      total_width = widths[key] + 1
      padding = " " * max(0, total_width - len(value))
      segments.append("| " + value + padding)
    data_row_list.append("".join(segments) + "|")
  data_rows = "\n".join(data_row_list)

  return "\n".join([
    border_row,
    header_row,
    border_row,
    data_rows,
    border_row
  ])